File: src/main/java/io/github/nikpivkin/lifetime/events/PluginStateChangedEvent.java
package io.github.nikpivkin.lifetime.events;

import org.bukkit.event.Event;
import org.bukkit.event.HandlerList;
import org.jetbrains.annotations.NotNull;

public abstract sealed class PluginStateChangedEvent extends Event {

  private static final HandlerList HANDLERS = new HandlerList();
  private final boolean enabled;

  protected PluginStateChangedEvent(boolean enabled) {
    this.enabled = enabled;
  }

  public static HandlerList getHandlerList() {
    return HANDLERS;
  }

  @Override
  public @NotNull HandlerList getHandlers() {
    return HANDLERS;
  }

  public boolean isEnabled() {
    return enabled;
  }

  public static final class Enabled extends PluginStateChangedEvent {

    public Enabled() {
      super(true);
    }
  }

  public static final class Disabled extends PluginStateChangedEvent {

    public Disabled() {
      super(false);
    }
  }
}

File: src/main/kotlin/io/github/nikpivkin/lifetime/events/PluginStateChangedEvent.kt
package io.github.nikpivkin.lifetime.events

import org.bukkit.event.Event
import org.bukkit.event.HandlerList

sealed class PluginStateChangedEvent : Event() {

    override fun getHandlers() = HANDLERS

    companion object {
        private val HANDLERS = HandlerList()

        @JvmStatic
        fun getHandlerList() = HANDLERS
    }

    class Enabled : PluginStateChangedEvent()
    class Disabled : PluginStateChangedEvent()
}
